source/common/filesystem/posix/filesystem_impl.cc

```diff
@@ -28,17 +28,38 @@ FileImplPosix::~FileImplPosix() {
   }
 }
 
-void FileImplPosix::openFile() {
-  const int flags = O_RDWR | O_APPEND | O_CREAT;
-  const int mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
-
-  fd_ = ::open(path_.c_str(), flags, mode);
+void FileImplPosix::openFile(FlagSet in) {
+  const auto flags_and_perm = translateFlag(in);
+  fd_ = ::open(path_.c_str(), flags_and_perm.flags, flags_and_perm.perm);
 }
 
 ssize_t FileImplPosix::writeFile(absl::string_view buffer) {
   return ::write(fd_, buffer.data(), buffer.size());
 }
 
+FileImplPosix::FlagsAndPerm FileImplPosix::translateFlag(FlagSet in) {
+  int out = 0;
+  int mode = 0;
+  if (in.test(File::Operation::Create)) {
+    out |= O_CREAT;
+    mode |= S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
+  }
+
+  if (in.test(File::Operation::Append)) {
+    out |= O_APPEND;
+  }
+
+  if (in.test(File::Operation::Read) && in.test(File::Operation::Write)) {
+    out |= O_RDWR;
+  } else if (in.test(File::Operation::Read)) {
+    out |= O_RDONLY;
+  } else if (in.test(File::Operation::Write)) {
+    out |= O_WRONLY;
```

Comment:
it looks like there is a prioritization semantic here which is why it is working for you to specify all the flags (which seems contradictory to me).

Maybe we can just specify the winning flags in the access logger?

Reply:
I'm not sure what mean of the `Maybe we can just specify the winning flags in the access logger?`

The prioritization is because 

> In other words, the combination O_RDONLY | O_WRONLY is a logical error, and certainly does not have the same meaning as O_RDWR.

Reply:
OK got it; this change lgtm now. Sorry for the misunderstanding.